### PR TITLE
Add validation to questions

### PR DESCRIPTION
**Pull request by WilliamMcCumstie**
The questions are now validated using a modified version of `JSON:Schema`. The validation struct looks like:

```
 validate:
   type:
   required:
   enum:
   minimum:
   maximum:
   exclusive_minimum:
   exclusive_maximum:
   items:
```

The `type` and `enum` fields are the same as `JSON:SCHEMA` and support `string`/`integer`/`number`/`boolean`/`array`.

The `mininmum`/`maximum`/`exclusive_minimum`/`exclusive_maximum` fields are the same as `JSON:SCHEMA` but use `snake_case` instead of `camalCase`. They can only be used with `number`/`integer` types.

The `required` field is equivalent to `ActiveSupport` `blank?` for strings, otherwise it is a not nil validation.

The `items` key can only be used with the `array` type. The `items` use mostly the same validation as the top level `validate` key. However arrays of arrays are not supported.

---

## Validate vs Format

As the above specifications are specified within the template's metadata, it itself is validated using `JSON:Schema`. There is a degree of coupling with the `format` key, however they achieve different purposes:

* `format` dictates how the question is presented to the user and if the value needs coercing,
* `validate` determines if the response value is correct.

There is no cross validation between `format` and `validate`. This means it is possible to write an "un-askable" question. A common example of this is `format: text` but `validate: integer`. Attempting the cross validate is tricky, as it's not immediately obvious how `select`/`multi-select` type questions should be handled.

Instead, a `number` format has been added which currently supports integers.

---

## Preforming the validation

After loading the `validate` schema, it is transliterated into a `JSON:Schema` complient specification. E.g.:

```
# Original Spec
    validate:
      type: 'string'
      required: true

# JSON:Schema
      "validate": {                                                                                                                                                                                                                                                              
        "allOf": [                                                                                                                      
          {                                                         
            "type": [                                               
              "string",                                             
              "null"                                                
            ],                                                      
            "title": "type",                                        
            "description": "Must be a string"                       
          },                                                        
          {                                                         
            "if": {                                                 
              "type": "null"                                        
            },                                                      
            "then": {                                               
              "type": "string",                                     
              "title": "required",                                  
              "description": "Must not be null"                     
            }                                                       
          },                                                        
          {                                                         
            "pattern": "^.+$",                                      
            "title": "required",                                    
            "description": "Must not be an empty string"            
          }                                                         
        ]                                                           
      }
```

An error message is generated when the spec is transliterated and stored as the `description`. These messages will eventually become part of the user facing error.

On `create-script`, the following will happen:
* When prompting, it will display the error and prompt again,
* With `--answers`, it validates the entire struct and raises the message.

---

The `format`/`default` validation has been updated to support `number`/"integer" types. The `items` key is now also validated for `select`/`multiselect` type questions.

### CODE REVIEW

**Comment by WilliamMcCumstie** [comment]
Rebased on master

**Review by benarmston — CHANGES_REQUESTED** [review]
This works well.  I've not been able to break it despite some effort to do so.  The error messages provided to the user are clear for both interactive and `--answers` usage.

There are a few areas that I'd like you to look at before I merge this.

Logging of JSON schema errors is much better than it was before.  However, The logging of invalid templates can still quite excessive.  If I remove the `version` field from a template, the entire template expressed as JSON ends up in the logs at warn level.  We don't need to include the template for top-level errors.  Is it possible to remove it?


A few of the files are getting far too large.  In particular the `commands/create_script.rb` and `models/template.rb`.  I find navigating large files rather annoyting.  It looks like the `QuestionPrompter` should be simple to extract to its own file.  Can you look at doing that please?  The other files can wait for another time.

The naming isn't quite right in a few places.  Classes should typically be named as nouns rather than verbs.  So `JSONSchemaErrorsLogger` in preference to `LogJSONSchemaErrors`.

Similar to the above, instances are often better named as nouns rather than verbs.  So instead of

```
@validate_value ||= JSONSchemer.schema(validate_schema)
@validate_value.validate(value)
```

it would be better to have

```
@answer_validator ||= JSONSchemer.schema(validate_schema)
@answer_validator.validate(answer)
```

---

There are also a few areas that need to be looked at in a future piece of work.

* If the `--json` flag is given, the errors should be returned as a JSON object.
* We should support number inputs other than integers.

File: lib/flight_job/commands/create_script.rb [hunk 1]
```diff
@@ -331,9 +354,55 @@ def prompt_question(question)
               q.default default
               q.validate(/\A24:00|([0-1]\d|2[0-3]):[0-5]\d\Z/, "Times must be in HH:MM format")
             end
+          when 'number'
+            # NOTE: The 'number' input has parity with HTML <input type="number"/>
+            # By default, this only allows integers. This behaviour has been replicated here
+            #
+            # Consider refactoring to allow floating points
```
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply lib/flight_job/commands/create_script.rb:361]: :+1:

File: lib/flight_job/models/question.rb [hunk 2]
```diff
@@ -27,11 +27,144 @@
...
     def related_question_id
       return nil unless ask_when
       ask_when['value'].split('.')[1]
     end
+
+    def required?
+      validate['required']
+    end
+
+    def validate_value(value)
```
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply lib/flight_job/models/question.rb]: 88002bb733b7e3f6a5d69a2102bf9fbe2ad1ac50

**Pull request merged**